File: Java/src/main/java/io/github/mihaistreames/afe/algorithms/sorting/QuickSort.java
package io.github.mihaistreames.afe.algorithms.sorting;

import org.jetbrains.annotations.NotNull;

import java.util.Comparator;
import java.util.List;

/**
 * An efficient, in-place, comparison-based sorting algorithm.
 * It is a divide and conquer algorithm.
 * <p>
 * This implementation is not stable. <br>
 * Time Complexity: O(n log n) average, O(n^2) worst-case <br>
 * Space Complexity: O(log n)
 */
public final class QuickSort {
    private QuickSort() {
        throw new UnsupportedOperationException("Utility class cannot be instantiated");
    }

    /**
     * Sorts a list using the quicksort algorithm and natural ordering.
     *
     * @param <T>  the type of elements, must extend {@link Comparable}
     * @param list the list to be sorted
     */
    public static <T extends Comparable<T>> void quickSort(@NotNull final List<T> list) {
        sort(list, 0, list.size() - 1, Comparator.naturalOrder());
    }

    /**
     * Sorts a list using the quicksort algorithm and a custom comparator.
     *
     * @param <T>        the type of elements
     * @param list       the list to be sorted
     * @param comparator the comparator to determine the order
     */
    public static <T> void quickSort(@NotNull final List<T> list,
                                     @NotNull final Comparator<T> comparator) {
        sort(list, 0, list.size() - 1, comparator);
    }

    private static <T> void sort(@NotNull final List<T> list,
                                 int low,
                                 int high,
                                 @NotNull final Comparator<T> comparator) {
        if (low < high) {
            int pi = partition(list, low, high, comparator);
            sort(list, low, pi - 1, comparator);
            sort(list, pi + 1, high, comparator);
        }
    }

    private static <T> int partition(@NotNull final List<T> list,
                                     int low,
                                     int high,
                                     @NotNull final Comparator<T> comparator) {
        final T pivot = list.get(high);
        int i = (low - 1);
        for (int j = low; j < high; j++) {
            if (comparator.compare(list.get(j), pivot) < 0) {
                i++;
                final T temp = list.get(i);
                list.set(i, list.get(j));
                list.set(j, temp);
            }
        }
        final T temp = list.get(i + 1);
        list.set(i + 1, list.get(high));
        list.set(high, temp);
        return i + 1;
    }
}

File: Python/afe/algorithms/sorting/quick_sort.py
# -*- coding: utf-8 -*-

from typing import List, TypeVar, Optional, Callable
from .sort_interface import SortInterface
import random


T = TypeVar('T')


class QuickSort(SortInterface):

    """
    A class implementing the Quick Sort algorithm. \n
    Quick Sort is a divide-and-conquer algorithm that selects a 'pivot' element and partitions the other elements
    into two sub-arrays according to whether they are less than or greater than the pivot. \n
    It is not a stable sort, meaning that it does not maintain the relative order of records with equal keys.
    """

    @staticmethod
    def _partition(data: List[T], low: int, high: int, comparator: Callable[[T, T], int]) -> int:
        # Use random pivot to avoid worst-case
        pivot_index = random.randint(low, high)
        data[pivot_index], data[high] = data[high], data[pivot_index]
        pivot = data[high]

        i = low - 1
        for j in range(low, high):
            if comparator(data[j], pivot) <= 0:
                i += 1
                data[i], data[j] = data[j], data[i]

        data[i + 1], data[high] = data[high], data[i + 1]
        return i + 1

    @staticmethod
    def _quick_sort_recursive(data: List[T], low: int, high: int, comparator: Callable[[T, T], int]):
        if low < high:
            pi = QuickSort._partition(data, low, high, comparator)
            QuickSort._quick_sort_recursive(data, low, pi - 1, comparator)
            QuickSort._quick_sort_recursive(data, pi + 1, high, comparator)

    @staticmethod
    def sort(data: List[T], comparator: Optional[Callable[[T, T], int]] = None) -> None:
        
        """Sorts a list in-place using the Quick Sort algorithm.

        This implementation is not stable. It uses a randomized pivot to achieve
        O(n log n) average time complexity.

        Args:
            data: The list of elements to sort.
            comparator: A function for comparing two elements. If None, default
                comparison operators are used.
        """

        if not data: return
        if comparator is None: comparator = QuickSort.DEFAULT_COMPARATOR

        QuickSort._quick_sort_recursive(data, 0, len(data) - 1, comparator)

